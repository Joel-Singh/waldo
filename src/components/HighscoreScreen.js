import { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { getTopTenHighscores, addHighscore } from "../util/firebase";

// eslint-disable-next-line
import highscoreScreenCss from "../styles/highscoreScreen.css";

export default function HighscoreScreen({
  map,
  currentPlayerScore = 99,
  backToSelectionScreenBtnOnClick = () => {},
}) {
  const { topTenHighscores, updateTopTenHighscores } =
    useTopTenHighscoresFromDatabase();
  const [scoreInputIsHidden, setScoreInputIsHidden] = useState(false);

  return (
    <div data-testid="HighscoreScreen" className="highscoreScreen">
      <ol
        data-testid="HighscoreScreen__scores-container"
        className="highscoreScreen__scores-container"
      >
        {topTenHighscores.map(({ initials, timeTaken }, index) => (
          <li key={index} data-position={index + 1}>
            <span>{initials}</span>
            <span>{timeTaken.toFixed(1)}</span>
          </li>
        ))}
      </ol>
      <div className="HighscoreScreen__player-score">
        Your score is: {currentPlayerScore.toFixed(1)} seconds
      </div>

      <ScoreInput />

      <Link
        to={"/"}
        className="highscoreScreen__back-btn"
      >
        <button
          onClick={backToSelectionScreenBtnOnClick}
          type="button"
        >
          Go back to selection screen
        </button>
      </Link>
    </div>
  );

  function ScoreInput() {
    const onUploadScoreClick = async () => {
      const initials = document.getElementById("initials").value;
      if (initials !== "") {
        setScoreInputIsHidden(true);
        await addHighscore(map, initials, currentPlayerScore);
        updateTopTenHighscores();
      }
    };

    return isCurrentPlayerScoreInTopTen() && !scoreInputIsHidden ? (
      <div className="highscoreScreen__score-input">
        <label htmlFor="initials">Enter Initials to upload score</label>
        <input id="initials" type="text" />
        <button
          type="button"
          onClick={onUploadScoreClick}
          className="highscoreScreen__score-input__upload-score-btn"
        >
          Upload score
        </button>
      </div>
    ) : null;

    function isCurrentPlayerScoreInTopTen() {
      let currentPlayerScoreIsInTopTen = false;
      if (topTenHighscores[9] !== undefined)
        currentPlayerScoreIsInTopTen =
          currentPlayerScore < topTenHighscores[9].timeTaken;

      return currentPlayerScoreIsInTopTen;
    }
  }

  function useTopTenHighscoresFromDatabase() {
    const [topTenHighscores, setTopTenHighscores] = useState([]);
    const [updateTopTenHighscoresState, setUpdateTopTenHighscoresState] =
      useState(0);

    useEffect(() => {
      setTopTenHighscores([]);
      let ignore = false;

      getTopTenHighscores(map).then((topTenHighscores) => {
        if (!ignore) setTopTenHighscores(topTenHighscores);
      });

      return () => {
        ignore = true;
      };
    }, [updateTopTenHighscoresState]);

    const updateTopTenHighscores = () =>
      setUpdateTopTenHighscoresState((prev) => prev + 1);
    return { topTenHighscores, updateTopTenHighscores };
  }
}

HighscoreScreen.propTypes = {
  map: PropTypes.string.isRequired,
};
