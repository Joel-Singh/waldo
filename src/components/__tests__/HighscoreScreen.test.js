import { act, getByTestId, render, screen, waitFor } from "@testing-library/react";
import getFirebaseFunctions from "../../util/firebase";
import { BrowserRouter } from "react-router-dom"
import HighscoreScreen from "../HighscoreScreen";

async function renderHighscoreScreen(highscoreScreen) {
  render(
    <BrowserRouter>
      {highscoreScreen}
    </BrowserRouter>
  );

  // need to wait for the initial fetch of the scores
  await act(async () => {
    await new Promise(resolve => setTimeout(resolve, 50))
  })

  return screen.getByTestId("HighscoreScreen")
}

test("Highscore screen renders", async () => {
  const highscoreScreen = await renderHighscoreScreen(<HighscoreScreen map="maze" />);
  expect(highscoreScreen).toBeDefined();
});

test("Highscore screen renders highscores", async () => {
  const { addHighscore, clearHighscores } = getFirebaseFunctions();

  const createHighscoreEntry = (initials, timeTaken) => ({ initials, timeTaken });
  const initialsAndScores = [
    createHighscoreEntry("AB", 1),
    createHighscoreEntry("CD", 2),
    createHighscoreEntry("EF", 3),
    createHighscoreEntry("GH", 4),
    createHighscoreEntry("IJ", 5),
    createHighscoreEntry("KL", 6),
    createHighscoreEntry("MN", 7),
    createHighscoreEntry("OP", 8),
    createHighscoreEntry("QR", 9),
    createHighscoreEntry("ST", 10),
  ];

  await clearHighscores()
  await Promise.all(
    initialsAndScores.map(({ initials, timeTaken }) =>
      addHighscore("maze", initials, timeTaken)
    )
  );

  const highscoreScreen = await renderHighscoreScreen(<HighscoreScreen map="maze" />);

  expect(highscoreScreen).toMatchSnapshot()
});

test("Highscore screen renders current score", async () => {
  await renderHighscoreScreen(<HighscoreScreen map="maze" currentPlayerScore={11.0} />)
  expect(screen.getByText('11.0', {exact: false})).toBeDefined()
})

it("doesn't offer the option to upload score when score isn't in top ten", async () => {
  const { addHighscore, clearHighscores } = getFirebaseFunctions();
  const createHighscoreEntry = (initials, timeTaken) => ({ initials, timeTaken });

  await clearHighscores()
  const initialsAndScores = [
    createHighscoreEntry("AB", 1),
    createHighscoreEntry("CD", 2),
    createHighscoreEntry("EF", 3),
    createHighscoreEntry("GH", 4),
    createHighscoreEntry("IJ", 5),
    createHighscoreEntry("KL", 6),
    createHighscoreEntry("MN", 7),
    createHighscoreEntry("OP", 8),
    createHighscoreEntry("QR", 9),
    createHighscoreEntry("ST", 10),
  ];

  await Promise.all(
    initialsAndScores.map(({ initials, timeTaken }) =>
      addHighscore("maze", initials, timeTaken)
    )
  );

  await renderHighscoreScreen(<HighscoreScreen map="maze" currentPlayerScore={11.0} />);
  expect(screen.queryByRole('textbox')).toBeNull()
})

it("does offer the option to upload score when score is in top ten", async () => {
  const { addHighscore, clearHighscores } = getFirebaseFunctions();
  const createHighscoreEntry = (initials, timeTaken) => ({ initials, timeTaken });

  await clearHighscores()
  const initialsAndScores = [
    createHighscoreEntry("AB", 1),
    createHighscoreEntry("CD", 2),
    createHighscoreEntry("EF", 3),
    createHighscoreEntry("GH", 4),
    createHighscoreEntry("IJ", 5),
    createHighscoreEntry("KL", 6),
    createHighscoreEntry("MN", 7),
    createHighscoreEntry("OP", 8),
    createHighscoreEntry("QR", 9),
    createHighscoreEntry("ST", 10),
  ];

  await Promise.all(
    initialsAndScores.map(({ initials, timeTaken }) =>
      addHighscore("maze", initials, timeTaken)
    )
  );

  await renderHighscoreScreen(<HighscoreScreen map="maze" currentPlayerScore={0.5} />);
  expect(screen.queryByRole('textbox')).not.toBeNull()
})
