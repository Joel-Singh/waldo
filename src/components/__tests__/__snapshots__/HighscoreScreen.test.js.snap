// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highscore screen renders highscores 1`] = `
<div
  class="highscoreScreen"
  data-testid="HighscoreScreen"
>
  <div
    class="highscoreScreen__scores-container"
    data-testid="HighscoreScreen__scores-container"
  >
    <div>
      <span>
        AB
      </span>
      <span>
        1
      </span>
    </div>
    <div>
      <span>
        CD
      </span>
      <span>
        2
      </span>
    </div>
    <div>
      <span>
        EF
      </span>
      <span>
        3
      </span>
    </div>
    <div>
      <span>
        GH
      </span>
      <span>
        4
      </span>
    </div>
    <div>
      <span>
        IJ
      </span>
      <span>
        5
      </span>
    </div>
    <div>
      <span>
        KL
      </span>
      <span>
        6
      </span>
    </div>
    <div>
      <span>
        MN
      </span>
      <span>
        7
      </span>
    </div>
    <div>
      <span>
        OP
      </span>
      <span>
        8
      </span>
    </div>
    <div>
      <span>
        QR
      </span>
      <span>
        9
      </span>
    </div>
    <div>
      <span>
        ST
      </span>
      <span>
        10
      </span>
    </div>
  </div>
  <div
    class="HighscoreScreen__player-score"
  >
    Your score is: 
    99.0
  </div>
  <a
    href="/"
  >
    <button
      class="highscoreScreen__back-btn"
      type="button"
    >
      Go Back to selection screen
    </button>
  </a>
</div>
`;
